Dev

v-nvtsk opened the pull request:
# OTUS JavaScript Developer. Basic | OTUS | Задание к занятию 24 (ДЗ 11)

https://otus.ru/learning/233931/

# Описание/Пошаговая инструкция выполнения домашнего задания:

## Выполнено

- [x] настроен линтинг и actions, автодеплой из PR
- [x] сборка webpack
- [x] добавить поддержку ts файлов через babel
- [x] добавить поддержку импорта css файлов
- [x] реализовать приложение "Игра Жизнь" на языке Typescript [Wiki](https://ru.wikipedia.org/wiki/%D0%98%D0%B3%D1%80%D0%B0_%C2%AB%D0%96%D0%B8%D0%B7%D0%BD%D1%8C%C2%BB)
- [x] ссылку на задеплоенную страницу и на пуллреквест сбросить в чат по дз
- [x] настроить jest и написать тесты на приложение

## Критерии оценки:

Игра должна поддерживать отображение нескольких независимых полей на странице (см пример)  
Критерии оценки

- [x] настроена поддержка typescript через babel - 5
- [x] реализовано приложение "Игра Жизнь" - 5 баллов
- [x] реализовано взаимодействие с полем (клик по ячейке меняет ее состояние, можно рисовать с зажатой левой кнопкой мыши) - 2
- [x] реализован автостоп игры, когда все клетки умерли - 2
- [x] реализован механизм изменения размеров поля (input поле (type=number)), в тч на лету (при увеличении заполнение мертвыми клетками, при уменьшении просто уничтожение ячеек) - 2
- [x] реализован механизм изменения скорости игры (input type=range) - 2
- [x] реализована подсветка клеток, которые являясь живыми должны умереть в следующем поколении (мертвые - чёрный цвет, живые - зелёный, обреченные на смерть - грязно-зелёный) - 2

v-nvtsk committed:
ci: add github actions

v-nvtsk committed:
build: add config files

v-nvtsk committed:
feat: add simple Life game

v-nvtsk committed:
feat: add coloring for dying cells

v-nvtsk committed:
refactor: change css colors declarations

v-nvtsk committed:
refactor: optimize render-grid

v-nvtsk committed:
fix: update github action for lint

v-nvtsk committed:
fix: update github action for prod build

v-nvtsk committed:
fix: update github action folder name for build

vvscode commented on .eslintrc:
```diff
@@ -0,0 +1,35 @@
+      }
+    ]
+  },
+  "ignorePatterns": ["node_modules", ".git", "*.config.ts"]
```
а чем конфиги не угодили?

luckfinder05 replied:
При проверке линтером *webpack.config.js* или *jest.config.ts* получал сообщения об ошибке:
```bash
  0:0  error  Parsing error: ESLint was configured to run on `<tsconfigRootDir>/coverage\lcov-report\block-navigation.js` using `parserOptions.project`: <tsconfigRootDir>/tsconfig.json      
However, that TSConfig does not include this file. Either:
- Change ESLint's list of included files to not include this file
- Change that TSConfig to include this file
- Create a new TSConfig that includes this file and include it in your parserOptions.project   
See the typescript-eslint docs for more info: https://typescript-eslint.io/linting/troubleshooting#i-get-errors-telling-me-eslint-was-configured-to-run--however-that-tsconfig-does-not--none-of-those-tsconfigs-include-this-file
```
Решения не нашёл (вернее по ссылке как раз и предлагают добавить в игнор), поэтому добавил, а когда появились другие проблемы (при проверке файлов в папках coverage, dist) прописал в скрипте путь для проверки только в src.
Оставил, чтобы IDE не показывала постоянно сообщения об ошибках.
Это неправильно?

vvscode replied:
тут тебе предлагают несколько решений на выбор, и мне не ясно, зачем ты выбрал это (поэтому и спрашиваю, чем конфиг не угодил)

можно же настроить, чтобы TS включал этот файл

luckfinder05 replied:
Попробую

vvscode commented on .prettierrc:
```diff
@@ -0,0 +1,11 @@
+{
+  "printWidth": 120,
+  "tabWidth": 2,
+  "semi": false,
```
вообще идея prettier как раз в том, чтобы его не настраивать

ну ок, `printWidth` это полезно - а остальное зачем?

vvscode commented on package.json:
```diff
@@ -0,0 +1,63 @@
+    "test": "jest ./src --coverage",
+    "test:watch": "jest ./src --watch",
+    "prepare": "husky install",
+    "lint-staged": "lint-staged",
```
а смысл? 

что это дает? (`npx lint-staged` такой же по длинне)

vvscode commented on src/append-with-element.ts:
```diff
@@ -0,0 +1,9 @@
+  return element
+}
+
+export default appendWithElement
```
https://habr.com/ru/articles/727542/

там не только про реакт

luckfinder05 replied:
Заменил все дефолт экспорты на именованные.
В функциях с большим количеством параметров передаю их в объекте.

vvscode commented on src/count-neighbours.ts:
```diff
@@ -0,0 +1,13 @@
+    for (let j = -1; j <= 1; j++) {
+      if (x + i >= 0 && y + j >= 0 && x + i <= field.length - 1 && y + j <= field.length - 1) {
+        const add = field[x + i][y + j]
+        count += add === 3 ? 1 : add
```
зачем так? те да, работает - но тебе самому это легко читается?

всякие https://eslint.org/docs/latest/rules/no-plusplus не просто так придумывают (а ты тут еще докручиваешь) - это вкусовщина, но ты проверь на одногрупниках, на сколько им эта запись заходит

программист сильно больше **читает** код, чем пишет

luckfinder05 replied:
честно говоря я и сам через день уже с трудом могу понять что там написано.
попробовал исправить

vvscode commented on src/create-grid.test.ts:
```diff
@@ -0,0 +1,36 @@
+  })
+
+  afterEach(() => {
+    document.body.innerHTML = ''
```
на занятии обсуждали, что вариант работы с глобальным ДОМ так себе, просто напоминаю

luckfinder05 replied:
Но мы же вроде остановились на том, что document.body.innerHTML надо сбрасывать перед каждым тестом?
Или лучше создавать новый элемент и в нём работать?

vvscode replied:
мы остановились, что если уж вы работаете с глобальным документом, то нужно сбрасывать (это как если ты себе отрезаешь ноги, то лучше промывать место отреза и бинтовать, но еще лучше вообще ноги не трогать)

vvscode commented on src/create-grid.ts:
```diff
@@ -0,0 +1,20 @@
+      cell.setAttribute('data-state', '0')
+      cell.classList.add('cell')
+      cell.style.cssText = `width: calc(100% / ${size}); height: calc(100% / ${size});`
+      row.appendChild(cell)
```
а можно сделать что-то вроде https://stackoverflow.com/questions/494143/creating-a-new-dom-element-from-an-html-string-using-built-in-dom-methods-or-pro и увеличить читабельность кода

luckfinder05 replied:
Применил. Но работает сильно медленнее createElement, поэтому не везде

vvscode replied:
"сильно медленнее" это в чем меряется? (я не спорю, просто спрашиваю про метрики)
может там можно просто сделать другие оптимизации, которые не влияют на читабельность

luckfinder05 replied:
тесты стали раза в 1.5-2 больше времени занимать

vvscode commented on src/create-markup.test.ts:
```diff
@@ -0,0 +1,61 @@
+
+  it('sizeInput should be initalized', () => {
+    expect(sizeInput).not.toBeNull()
+    if (sizeInput !== null) {
```
https://stackoverflow.com/questions/27880631/why-should-you-avoid-conditional-logic-in-unit-tests-and-how

vvscode replied:
зачем тут `if` если ты точно знаешь, что он не `.null` (ты это уже проверил)

luckfinder05 replied:
Typescript требует...
![image](https://github.com/v-nvtsk/otus-jsbasic-dz24-life/assets/138329268/02e2fa38-ab1d-43d9-8914-53aff17b2366)

Тут у меня, кстати, вопрос - почти везде пришлось сначала прописывать такие if, а потом заметил, что некоторые из них можно безболезненно удалить, но не этот.
Из-за чего это и как избежать повторение этого?
Можно ли использовать "!"? сейчас он запрещен правилами линтера

vvscode replied:
ну он требует ( и правильно ), но его требования можно по-разному разрешить

например, использовать `as XXX` (сужение типа, чтобы убрать `| null`) или

```
if (!sizeInput) {
 throw new Error('not found')
}
```
тут если элемента не будет тест упадет

vvscode commented on src/game.ts:
```diff
@@ -0,0 +1,137 @@
+  constructor(container: HTMLElement, size: number = 30, timeInterval: number = 400) {
+    this.container = container
+    this.timeInterval = timeInterval
+    this.size = size
```
https://www.typescriptlang.org/docs/handbook/2/classes.html#parameter-properties

luckfinder05 replied:
Применил

vvscode commented on src/game.ts:
```diff
@@ -0,0 +1,137 @@
+      },
+    }
+
+    const mouseCallbacks = {
```
1) какие причины держать это в конструкторе? https://stackoverflow.com/questions/7048515/is-doing-a-lot-in-constructors-bad
2) какие причины не вынести функции к методы класса (этого или другого, чтобы не создавать каждый раз + чтобы повысить читабельность)

vvscode replied:
3 https://stackoverflow.com/questions/2976834/how-big-should-function-length-be-lines-of-code-in-a-function

luckfinder05 replied:
Если вынести в метод init - это нормально или имеется ввиду что-то другое?
Слишком там всё связано друг с другом, чтобы разделять.

vvscode replied:
если вынести в метод init - это закроет п1 (просто перенесет в другой метод)

но ты сейчас сам говоришь, что у тебя слишком связанный код (менять его по частям ты не можешь) что плохо - может стоит подумать про улучшение?

vvscode commented on src/index.ts:
```diff
@@ -0,0 +1,15 @@
+const game = new Game(container)
+
+// const container1 = document.createElement('div')
+// container1.className = 'game-container'
```
давай все же чистить код перед сдачей

luckfinder05 replied:
Конечно

vvscode commented on src/render-grid.ts:
```diff
@@ -0,0 +1,74 @@
+// import initArray from './init-array'
+
+export default function renderGrid(container: HTMLElement, size: number, data: number[][] | null): number {
```
я с все той же ссылкой https://stackoverflow.com/questions/2976834/how-big-should-function-length-be-lines-of-code-in-a-function

vvscode commented on src/render-grid.ts:
```diff
@@ -0,0 +1,74 @@
+        row.forEach((cell, j) => {
+          if (j < size) {
+            cell.setAttribute('data-state', gridData[i][j].toString())
+            cell.style.cssText = `width: calc(100% / ${size}); height: calc(100% / ${size});`
```
а почему не вынести все это в css? 
https://developer.mozilla.org/en-US/docs/Web/CSS/--*  в помощь

luckfinder05 replied:
сделал через css-var и её изменение в js

vvscode commented on src/store.ts:
```diff
@@ -0,0 +1,44 @@
+    this.setCells(newState)
+  }
+
+  public getCells(): number[][] {
```
а в чем смысл этого метода? в чем разница с публичным store?

luckfinder05 replied:
стор не публичный. хотелось ограничить обращения в него извне, метод добавил только для тестов

vvscode replied:
ну он не публичный, но твои сеттеры и геттеры делают его таким

http://blog.kislenko.net/show.php?id=1775 например почитай

v-nvtsk committed:
feat: add createElementFromHTML function

v-nvtsk committed:
feat: add createCell function

v-nvtsk committed:
refactor: countNeighbours

v-nvtsk committed:
refactor: remove append-with-element

v-nvtsk committed:
refactor: change after review

v-nvtsk committed:
style: update prettier config

v-nvtsk committed:
build: update eslint config, package.json

v-nvtsk committed:
fix:  remove commented code

v-nvtsk committed:
refactor: change params to parameter properties

vvscode commented on src/style.css:
```diff
@@ -0,0 +1,52 @@
+:root {
+  --grid-size: 30;
+}
+
```
кажется, что в папке src уже можно подпапки создавать

v-nvtsk committed:
feat: add debouncer for field resize

v-nvtsk committed:
refactor: move utility functions to utils dir

v-nvtsk committed:
refactor: add explanatory comments for renderGrid

v-nvtsk committed:
fix: set grid-size css-var on create markup

v-nvtsk committed:
refactor: rename var for consistent naming

v-nvtsk committed:
refactor game.ts...

move callbacks to separate methods
remove onStart, onStop excessive methods
add resize debouncer
remove getCells method

v-nvtsk committed:
feat: add touch support

v-nvtsk committed:
fix: prettier script string

v-nvtsk committed:
remove comments

v-nvtsk committed:
fix: add eslint ability to process root files

v-nvtsk committed:
fix: update package.json

change scripts to run prettier and lint from root

v-nvtsk committed:
style: update config formatting

v-nvtsk committed:
feat: add app title in webpack config

v-nvtsk committed:
feat: add README.md

v-nvtsk committed:
fix: webpack.config document title move to HTMLWebpackPlugin
